### PR TITLE
chore(docs): Add preliminary docs on build-and-publish pipeline.

### LINKED ISSUE #601: Document our CI setup and how to maintain it
(I'm doing a refresh of the docs setup in this repo and filing issues for things that I can't find, but want).

It's not obvious to me why we have all three of travis, circleci and taskcluster set up this repo, or which one I should look at for what reasons.

### PR DESCRIPTION
**Pull request by rfk**
Fixes #601.

Here's my first attempt to capturing the different moving parts in our build pipeline, and some brief comments on how we ensure end-to-end integrity of the builds. @eoger, how close did I come to actually documenting reality?

### Pull Request checklist ###

- [ ] ~~**Quality**: This PR builds and tests run cleanly~~
- [ ] ~~**Tests**: This PR includes thorough tests or an explanation of why it does not~~
- [ ] ~~**Changelog**: This PR includes a changelog entry or an explanation of why it does not need one~~

This is a docs-only change.

### CODE REVIEW

File: docs/build-and-publish-pipeline.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+## Application Services Build and Publish Pipeline
```
Comment (vladikoff) [inline docs/build-and-publish-pipeline.md]: Could we make a table of contents or an index of docs under https://github.com/mozilla/application-services/tree/master/docs/howtos ? I can never properly navigate the how-tos by looking at this view http://v14d.com/i/0ccf88a289f1336dcb04769576ad23ad60c1cedd4460ccce3d6d65c8a96693aa.jpg Am I doing it wrong?

File: docs/build-and-publish-pipeline.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+for development and maintenance purposes, and to serve as a basic analysis of the integrity
+protections that it offers (so you'll notice there are notes and open questions in place where
+we haven't fully hashed out all those details).
+
+The key points:
+
+* We use [Cargo](https://github.com/rust-lang/cargo) for building and testing the core Rust code in isolation,
+  [Gradle](https://gradle.org/) with [rust-android-gradle](https://github.com/mozilla/rust-android-gradle)
+  for combining Rust and Kotlin code into Android components and running tests against them,
+  and [Carthage](https://github.com/Carthage/Carthage) for combining Rust and Swift code into iOS components.
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Carthage doesn't do much really besides making distribution easier, the "combining" phase is all XCode.

File: docs/build-and-publish-pipeline.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+* [TaskCluster](../automation/taskcluster/README.md) runs on every pull-request, tag
+  and push to master, to ensure Android artifacts build correctly and to execute their
+  tests via gradle.
+* [CircleCI](../.circleci/config.yml) runs on every branch, pull-request, and tag,
+  to execute lint checks and automated tests at the rust level.
+    * TODO: how do we run automated tests for the Swift wrapper code?
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: https://github.com/mozilla/application-services/issues/607

File: docs/build-and-publish-pipeline.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+* Releases are made by [manually creating a new tag](./howtos/cut-a-new-release.md),
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Actually TC is not triggered by tags anymore but Github releases, it's pretty much similar as we create a tag at the same time but I'm not quite sure the release task will run on tags pushed using `git push --tags`.

File: docs/build-and-publish-pipeline.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+The key points:
+
+* We use [Cargo](https://github.com/rust-lang/cargo) for building and testing the core Rust code in isolation,
+  [Gradle](https://gradle.org/) with [rust-android-gradle](https://github.com/mozilla/rust-android-gradle)
+  for combining Rust and Kotlin code into Android components and running tests against them,
+  and [Carthage](https://github.com/Carthage/Carthage) for combining Rust and Swift code into iOS components.
+* [TaskCluster](../automation/taskcluster/README.md) runs on every pull-request, tag
+  and push to master, to ensure Android artifacts build correctly and to execute their
+  tests via gradle.
+* [CircleCI](../.circleci/config.yml) runs on every branch, pull-request, and tag,
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Ditto tags conversation from bellow

File: docs/build-and-publish-pipeline.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  to execute lint checks and automated tests at the rust level.
+    * TODO: how do we run automated tests for the Swift wrapper code?
+* Releases are made by [manually creating a new tag](./howtos/cut-a-new-release.md),
+  which triggers various CI jobs:
+    * [CircleCI](../.circleci/config.yml) is used to build an iOS binary release on every tag,
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: ditto tags

File: docs/build-and-publish-pipeline.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      and publish it as a GitHub release artifcact.
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Maybe we should also say that Carthage is always happy to use any release, but if a binary package is not present in the github release it'll build from source?

File: docs/build-and-publish-pipeline.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    * [TaskCluster](../automation/taskcluster/README.md) is used to:
+        * Build an Android binary release.
+        * Publish it to nalexander@'s personal bintray (although this will soon change to publish
+           to https://maven.mozilla.org).
+        * Upload symbols to [Socorro](https://wiki.mozilla.org/Socorro).
+           * TODO: does this mean the symbols are only valid for debugging Android crashes, not on iOS?
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Only Android lib symbols are generated.

File: docs/build-and-publish-pipeline.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+For Android consumers these are the steps by which Application Services code becomes available,
+and the integrity-protection mechanisms that apply at each step:
+
+1. Code is developed in branches and lands on `master` via pull request.
+    * GitHub branch protection prevents code being pushed to `master` without review.
+      * TODO: we should consider requiring signed tags before merge to `master`.
+    * CircleCI and TaskCluster run automated tests against the code, but do not have
+      the ability to push modified code back to GitHub.
+      * TODO: Or do they? Could we do more to guard against this?
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: I don't see how they could, they are just doing an https git clone with no keys/auth.

File: docs/build-and-publish-pipeline.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+3. TaskCluster checks out the tag, builds it for all target platforms, and runs automated tests.
+    * These tasks run in a pre-build docker image, helping assure integrity of the build environment.
+    * TODO: could this step check for signed tags as an additional integrity measure?
+4. TaskCluster uploads built artifacts to bintray
+    * Secret key for uploading to bintray is guaded by a TaskCluster scope that's only available
+      to this task.
+    * TODO: we're in the process of moving this to maven
+    * TODO: could a malicious dev dependency from step (3) influence the build environment here?
+    * TODO: talk about how TC's "chain of trust" might be useful here.
+5. Consumers fetch the published artifacts from bintray.
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Bintray artifacts are replicated to bintray's jcenter first, but it's more like a technicality (although at the moment broken!).

File: docs/build-and-publish-pipeline.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        * Upload symbols to [Socorro](https://wiki.mozilla.org/Socorro).
+           * TODO: does this mean the symbols are only valid for debugging Android crashes, not on iOS?
+
+For Android consumers these are the steps by which Application Services code becomes available,
+and the integrity-protection mechanisms that apply at each step:
+
+1. Code is developed in branches and lands on `master` via pull request.
+    * GitHub branch protection prevents code being pushed to `master` without review.
+      * TODO: we should consider requiring signed tags before merge to `master`.
+    * CircleCI and TaskCluster run automated tests against the code, but do not have
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Something we should talk about is how we disabled running TC tasks for the general public (only collaborators can trigger TC tasks with their PRs).

File: docs/build-and-publish-pipeline.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    * TODO: could a malicious dev dependency from step (3) influence the build environment here?
+    * TODO: talk about how TC's "chain of trust" might be useful here.
+5. Consumers fetch the published artifacts from bintray.
+
+For iOS consumers the corresponding steps are:
+
+1. Code is developed in branches and lands on `master` via pull request, as above.
+2. Developers manually create a release tag from latest `master`, as above.
+3. CircleCI checks out the tag, builds it, and runs automated tests.
+    * TODO: These tasks to `curl https://sh.rustup.rs | sh` and `pip install` and similar,
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: We should probably do the same as Taskcluster Android and grab `rustup-init` directly and check its sha256 checksum.

File: docs/build-and-publish-pipeline.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      is that unnecessarily risk?
+    * TODO: could this step check for signed tags as an additional integrity measure?
+4. CircleCI runs Carthage to assemble a zipfile of built frameworks.
+    * TODO: could a malicious dev dependency from step (3) influence the build environment here?
+5. CircleCI the [dpl](https://github.com/travis-ci/dpl) to publish to GitHub as a release artifact.
+    * CircleCI config contains a github token with appropriate permissions.
+    * TODO: who own this github token, and what permissions does it have?
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: This is mine, we should probably consider switching to something owned by Mozilla instead.

File: docs/build-and-publish-pipeline.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    * TODO: how do we run automated tests for the Swift wrapper code?
+* Releases are made by [manually creating a new tag](./howtos/cut-a-new-release.md),
+  which triggers various CI jobs:
+    * [CircleCI](../.circleci/config.yml) is used to build an iOS binary release on every tag,
+      and publish it as a GitHub release artifcact.
+    * [TaskCluster](../automation/taskcluster/README.md) is used to:
+        * Build an Android binary release.
+        * Publish it to nalexander@'s personal bintray (although this will soon change to publish
+           to https://maven.mozilla.org).
+        * Upload symbols to [Socorro](https://wiki.mozilla.org/Socorro).
```
Comment (eoger) [inline docs/build-and-publish-pipeline.md]: Not that the API key used to upload to Socorro is tied to my own LDAP account, apparently that's what they do.
Anyway if one day I go away you'd have to re-create that key.

File: docs/build-and-publish-pipeline.md (hunk 1 above)
Comment (rfk) [reply docs/build-and-publish-pipeline.md]: 👍, yes, excellent suggestion.

File: docs/build-and-publish-pipeline.md (hunk 9 above)
Comment (rfk) [reply docs/build-and-publish-pipeline.md]: One scenario: a compromised third-party dependency adds malicious code in its `build.rs` which reads `${GITHUB_TOKEN}` from the environment and uses it to make modifications to the main repo.  I don't *think* that the CircleCI workers have enough permissions to enable that kind of thing, but I'd like to understand for certain.

File: docs/build-and-publish-pipeline.md (hunk 13 above)
Comment (rfk) [reply docs/build-and-publish-pipeline.md]: Great, thanks, I'll file a follow-up. What permissions does this token currently have?

Comment (eoger) [reply docs/build-and-publish-pipeline.md]: `repo:public_repo`

Comment (rfk) [reply docs/build-and-publish-pipeline.md]: Per [1], this scope "Limits access to public repositories. That includes read/write access to code, commit statuses, repository projects, collaborators, and deployment statuses for public repositories and organizations."

Which seems like a lot, but github really doesn't do a good job of fine-grained access control :-(

Branch protection (and the recently-enabled requirement on signed commits) should in theory protect us from abuse of this token.

[1] https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/

**Comment by rfk** [comment]
Thanks @eoger, I've rebased and updated with your suggestions.

**Pull request merged**